Log exceptions from exceptionHandler method

sharidas opened the pull request:
Log exceptions from console.php

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
Log exceptions from console.php. As of now the exceptions are thrown in the console and then the script is exited. Before the script exits, it would be better to log the exception. This change set achieves the goal.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Log the exception that happens in the `exceptionHandler()` in console.php

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- I updated the changes as shown below
   - Before line `require_once 'public/Constants.php';` in lib/base.php add `hello`
   - Now run `./occ a:l`, you should see exception:
```
Cannot load Xdebug - it was already loaded
System is broken.
ParseError: syntax error, unexpected 'require_once' (T_REQUIRE_ONCE) in /home/sujith/test/owncloud3/lib/base.php:61
Stack trace:
#0 /home/sujith/test/owncloud3/occ(11): require_once()
#1 {main}%
```
   - Now remove `hello` and add `throw new \Exception('foo');` and re-run the command `./occ a:l`, you should see message like this in the console.
```
System is broken.
Exception: foo in /home/sujith/test/owncloud3/lib/base.php:60
Stack trace:
#0 /home/sujith/test/owncloud3/console.php(65): require_once()
#1 /home/sujith/test/owncloud3/occ(11): require_once('/home/sujith/te...')
#2 {main}%
```

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

sharidas committed:
Log exceptions from exceptionHandler method

Log exceptions from exceptionHandler method
in console.php.
Also reduce the size of try/catch block. And
restrict it to the require statement.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

jvillafanez commented:
I think this makes a bit more sense, assuming the exception handler logs things properly into the ownCloud log. (Need to soften the messages)
```
try {
	require_once __DIR__ . '/lib/base.php';
} catch (Exception $e) {
    echo "System broken beyond repair" . PHP_EOL;
    echo $e;
    exit(1);
} catch (Error $e) {
    echo "System broken beyond repair" . PHP_EOL;
    echo $e;
    exit(1);
}

// set to run indefinitely if needed
\set_time_limit(0);
if (!OC::$CLI) {
	echo 'This script can be run from the command line only' . PHP_EOL;
	exit(1);
}

\set_exception_handler(function($exception) {
        echo 'An unhandled exception has been thrown:' . PHP_EOL;
	echo $exception;
	\OC::$server->getLogger()->logException($exception);
	exit(1);
});
......
```

We have a try / catch block just for the `require_once`, in case something is really broken there. Note that the logger might not be available at that point, that's why I'd rather not rely on it.

For the rest of the code, we rely on the exception handler we're setting up. There is no need for a huge try / catch block when we already have set the exception handler the way we want. We might want to set the exception handler just below the "require_once".

In any case, we're relying on the logger to be working. I think we need to consider that the logger might be broken too. We might get an exception while getting the logger or while logging the exception inside the exception handler. These additional problems should be made visible.

sharidas commented:
Changes made:
- Reduced the try/catch block for `require_once __DIR__ . '/lib/base.php';`
- Added an echo for the `catch` -> `System is broken`. 
- Removed the exception in the try/catch block. The catch block will now echo and exit.

jvillafanez commented on console.php:
```diff
@@ -58,54 +58,59 @@
 try {
 	require_once __DIR__ . '/lib/base.php';
+} catch (Exception $ex) {
+	echo "System is broken." . PHP_EOL;
```
Any softer message we can use?

sharidas replied:
Let me know if this message looks ok?

jvillafanez replied:
Maybe just the "something went wrong" is enough. We'll show the stacktrace in the console, which can be seen as an addition to the message, so it might be fine. There won't be errors showing in the logs at that point, so leading the user to the log file doesn't look good.

Note that the message below (the one for the error) should also be adjusted

sharidas replied:
Adjusted the first message:
- "Something went wrong"
- second message as "Something fatal happened".
Let me know if this looks ok.

jvillafanez commented:
Note that except for a crash in the base.php file, the rest of unhandled exceptions should appear in the onwCloud log. Throwing a random exception in any console command should be enough to check this.

jvillafanez commented:
Please confirm that throwing a random exception in any command shows that exception in the logs. This is the main use case for this PR. It should work, but I don't see confirmation.

sharidas commented:
Test to verify logging of exception thrown from command:

- Modified the code at lib/private/Console/Application.php by adding:
```
diff --git a/lib/private/Console/Application.php b/lib/private/Console/Application.php
index cdccba2b5d..023ee72cd0 100644
--- a/lib/private/Console/Application.php
+++ b/lib/private/Console/Application.php
@@ -161,6 +161,7 @@ class Application {
                        ConsoleEvent::EVENT_RUN,
                        $args
                ));
+               $this->application->setCatchExceptions(false);
                return $this->application->run($input, $output);
        }
```

Now in the ListApps.php I added:
```
diff --git a/core/Command/App/ListApps.php b/core/Command/App/ListApps.php
index f3345e8b5f..575d6ec5cc 100644
--- a/core/Command/App/ListApps.php
+++ b/core/Command/App/ListApps.php
@@ -25,6 +25,7 @@
 namespace OC\Core\Command\App;
 
 use OC\Core\Command\Base;
+use OCP\App\AppNotFoundException;
 use OCP\App\IAppManager;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Input\InputArgument;
@@ -65,6 +66,7 @@ class ListApps extends Base {
        }
 
        protected function execute(InputInterface $input, OutputInterface $output) {
+               throw new AppNotFoundException("OOps I did this");
                $appNameSubString = $input->getArgument('search-pattern');
 
                if ($input->getOption('shipped') === 'true' || $input->getOption('shipped') === 'false') {
```

- Ran `./occ a:l`,  and here is the log obtained in log file:
```
{"reqId":"VmQSP2cXiAmEWbNiTdm8","level":3,"time":"2019-06-11T12:40:48+00:00","remoteAddr":"","user":"--","app":"no app in context","method":"--","url":"--","message":"Exception: {\"Exception\":\"OCP\\\\App\\\\AppNotFoundException\",\"Message\":\"OOps I did this\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Command\\\/Command.php(255): OC\\\\Core\\\\Command\\\\App\\\\ListApps->execute(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#1 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/core\\\/Command\\\/Base.php(159): Symfony\\\\Component\\\\Console\\\\Command\\\\Command->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#2 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(969): OC\\\\Core\\\\Command\\\\Base->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#3 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(255): Symfony\\\\Component\\\\Console\\\\Application->doRunCommand(Object(OC\\\\Core\\\\Command\\\\App\\\\ListApps), Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#4 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(148): Symfony\\\\Component\\\\Console\\\\Application->doRun(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#5 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/Console\\\/Application.php(165): Symfony\\\\Component\\\\Console\\\\Application->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#6 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/console.php(116): OC\\\\Console\\\\Application->run()\\n#7 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/occ(11): require_once('\\\/home\\\/sujith\\\/te...')\\n#8 {main}\",\"File\":\"\\\/home\\\/sujith\\\/test\\\/owncloud3\\\/core\\\/Command\\\/App\\\/ListApps.php\",\"Line\":69}"}
```

jvillafanez commented on console.php:
```diff
@@ -58,54 +58,59 @@
+	echo 'The process control (PCNTL) extensions are required in case you want to interrupt long running commands - see http://php.net/manual/en/book.pcntl.php' . PHP_EOL;
 }
+
+$application = new Application(\OC::$server->getConfig(), \OC::$server->getEventDispatcher(), \OC::$server->getRequest());
```
Looks like disabling we should disable the exception catching of the symfony's application. This should be the place for that because this is part of the setup, otherwise we won't be able to log the exceptions into the owncloud's log file.

sharidas replied:
Updated the PR.

sharidas commented:
Updated the PR and tested with the change set :
- run `./occ a:l` and user can see in the console:
```
sujith@sujith-ownCloud  ~/test/owncloud3   log-exception-console ●  ./occ a:l
Cannot load Xdebug - it was already loaded
An unhandled exception has been thrown:
OCP\App\AppNotFoundException: OOps I did this in /home/sujith/test/owncloud3/core/Command/App/ListApps.php:69
Stack trace:
#0 /home/sujith/test/owncloud3/lib/composer/symfony/console/Command/Command.php(255): OC\Core\Command\App\ListApps->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /home/sujith/test/owncloud3/core/Command/Base.php(159): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#2 /home/sujith/test/owncloud3/lib/composer/symfony/console/Application.php(969): OC\Core\Command\Base->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#3 /home/sujith/test/owncloud3/lib/composer/symfony/console/Application.php(255): Symfony\Component\Console\Application->doRunCommand(Object(OC\Core\Command\App\ListApps), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#4 /home/sujith/test/owncloud3/lib/composer/symfony/console/Application.php(148): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#5 /home/sujith/test/owncloud3/lib/private/Console/Application.php(163): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#6 /home/sujith/test/owncloud3/console.php(120): OC\Console\Application->run()
#7 /home/sujith/test/owncloud3/occ(11): require_once('/home/sujith/te...')
#8 {main}%                                                                                                                                                                                    ✘ sujith@sujith-ownCloud  ~/test/owncloud3   log-exception-console ● 
```
- Also in the log file following is logged:
```
{"reqId":"V1TZons2whIRBJNw5Ib4","level":3,"time":"2019-06-11T13:47:07+00:00","remoteAddr":"","user":"--","app":"no app in context","method":"--","url":"--","message":"Exception: {\"Exception\":\"OCP\\\\App\\\\AppNotFoundException\",\"Message\":\"OOps I did this\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Command\\\/Command.php(255): OC\\\\Core\\\\Command\\\\App\\\\ListApps->execute(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#1 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/core\\\/Command\\\/Base.php(159): Symfony\\\\Component\\\\Console\\\\Command\\\\Command->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#2 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(969): OC\\\\Core\\\\Command\\\\Base->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#3 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(255): Symfony\\\\Component\\\\Console\\\\Application->doRunCommand(Object(OC\\\\Core\\\\Command\\\\App\\\\ListApps), Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#4 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/composer\\\/symfony\\\/console\\\/Application.php(148): Symfony\\\\Component\\\\Console\\\\Application->doRun(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#5 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/Console\\\/Application.php(163): Symfony\\\\Component\\\\Console\\\\Application->run(Object(Symfony\\\\Component\\\\Console\\\\Input\\\\ArgvInput), Object(Symfony\\\\Component\\\\Console\\\\Output\\\\ConsoleOutput))\\n#6 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/console.php(120): OC\\\\Console\\\\Application->run()\\n#7 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/occ(11): require_once('\\\/home\\\/sujith\\\/te...')\\n#8 {main}\",\"File\":\"\\\/home\\\/sujith\\\/test\\\/owncloud3\\\/core\\\/Command\\\/App\\\/ListApps.php\",\"Line\":69}"}
```

jvillafanez commented:
Any chance to fetch the `Application` from the server? Something like
```
$application = \OC::$server->query(Application::class);
```
This should ensure we get the same instance created the same way across ownCloud.

sharidas commented:
> Any chance to fetch the `Application` from the server? Something like
> 
> ```
> $application = \OC::$server->query(Application::class);
> ```
> 

I wish it to be possible because, it fails to instantiate the class `EventDispatcherInterface` while trying to fetch `Application` from the server. Below is the error:
```
An unhandled exception has been thrown:
OCP\AppFramework\QueryException: Could not resolve Symfony\Component\EventDispatcher\EventDispatcherInterface! Class can not be instantiated in /home/sujith/test/owncloud3/lib/private/AppFramework/Utility/SimpleContainer.php:92
Stack trace:
#0 /home/sujith/test/owncloud3/lib/private/AppFramework/Utility/SimpleContainer.php(110): OC\AppFramework\Utility\SimpleContainer->resolve('Symfony\\Compone...')
#1 /home/sujith/test/owncloud3/lib/private/ServerContainer.php(86): OC\AppFramework\Utility\SimpleContainer->query('Symfony\\Compone...')
#2 /home/sujith/test/owncloud3/lib/private/AppFramework/Utility/SimpleContainer.php(64): OC\ServerContainer->query('Symfony\\Compone...')
#3 /home/sujith/test/owncloud3/lib/private/AppFramework/Utility/SimpleContainer.php(90): OC\AppFramework\Utility\SimpleContainer->buildClass(Object(ReflectionClass))
#4 /home/sujith/test/owncloud3/lib/private/AppFramework/Utility/SimpleContainer.php(110): OC\AppFramework\Utility\SimpleContainer->resolve('OC\\Console\\Appl...')
#5 /home/sujith/test/owncloud3/lib/private/ServerContainer.php(86): OC\AppFramework\Utility\SimpleContainer->query('OC\\Console\\Appl...')
#6 /home/sujith/test/owncloud3/console.php(118): OC\ServerContainer->query('OC\\Console\\Appl...')
#7 /home/sujith/test/owncloud3/occ(11): require_once('/home/sujith/te...')
#8 {main}{"reqId":"mKQCfG4BmdJXL8VdHKk7","level":3,"time":"2019-06-12T05:31:43+00:00","remoteAddr":"","user":"--","app":"no app in context","method":"--","url":"--","message":"Exception: {\"Exception\":\"OCP\\\\AppFramework\\\\QueryException\",\"Message\":\"Could not resolve Symfony\\\\Component\\\\EventDispatcher\\\\EventDispatcherInterface! Class can not be instantiated\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(110): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('Symfony\\\\\\\\Compone...')\\n#1 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/ServerContainer.php(86): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('Symfony\\\\\\\\Compone...')\\n#2 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(64): OC\\\\ServerContainer->query('Symfony\\\\\\\\Compone...')\\n#3 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(90): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->buildClass(Object(ReflectionClass))\\n#4 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(110): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('OC\\\\\\\\Console\\\\\\\\Appl...')\\n#5 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/ServerContainer.php(86): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('OC\\\\\\\\Console\\\\\\\\Appl...')\\n#6 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/console.php(118): OC\\\\ServerContainer->query('OC\\\\\\\\Console\\\\\\\\Appl...')\\n#7 \\\/home\\\/sujith\\\/test\\\/owncloud3\\\/occ(11): require_once('\\\/home\\\/sujith\\\/te...')\\n#8 {main}\",\"File\":\"\\\/home\\\/sujith\\\/test\\\/owncloud3\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php\",\"Line\":92}"}
```

We can by registering a service in Server.php. But imho, it would be too much for this. Let me know your opinion.

jvillafanez commented:
> We can by registering a service in Server.php. But imho, it would be too much for this. Let me know your opinion.

Let's leave it as it is for now.

There is an instance of the console application in "core/Controller/OccController.php". Could you find where it's being used and check if it still works? Other than that, I don't think we're missing anything.

sharidas commented:
> 
> There is an instance of the console application in "core/Controller/OccController.php". Could you find where it's being used and check if it still works? Other than that, I don't think we're missing anything.

I commented the validated by making a small change in the code though. I commented out https://github.com/owncloud/core/blob/master/core/Controller/OccController.php#L97 and executed:
```
sujith@sujith-ownCloud  ~/test/owncloud3   log-exception-console ●  curl -u admin:admin -X POST 'http://localhost/testing3/index.php/occ/status'    
{"exitCode":0,"response":"  - installed: true\n  - first_install_version: 11.0.0.0\n  - version: 11.0.0.0\n  - versionstring: 11.0.0 prealpha\n  - edition: Enterprise\n"}%                   
sujith@sujith-ownCloud  ~/test/owncloud3   log-exception-console ● 
```

So yes the execute method is working in the OccController.

jvillafanez commented on console.php:
```diff
@@ -58,54 +58,63 @@
+$symfonyApplication = new \Symfony\Component\Console\Application($defaults->getName(), \OC_Util::getVersionString());
+$symfonyApplication->setCatchExceptions(false);
+$application = new Application(\OC::$server->getConfig(), $symfonyApplication, \OC::$server->getEventDispatcher(), \OC::$server->getRequest());
+//$application->setCatchExceptions(false);
```
This isn't needed.

sharidas replied:
Done

jvillafanez commented on console.php:
```diff
@@ -58,54 +58,63 @@
+
+$defaults = new OC_Defaults();
+$symfonyApplication = new \Symfony\Component\Console\Application($defaults->getName(), \OC_Util::getVersionString());
+$symfonyApplication->setCatchExceptions(false);
```
Better include a comment here to know why we want this, otherwise we might remove this in the future.

sharidas replied:
Added comment.

jvillafanez commented:
ok, last touches and I think this is ready.

jvillafanez commented:
There are some tests failing, but these changes shouldn't be the cause....

sharidas committed:
Verify the stdout for the exceptions thrown for occ

Verify the stdout for the exceptions thrown for
occ command in the feature file.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

jvillafanez commented:
pinging @phil-davis or @individual-it for the change in the acceptance test, in case they want to add something (looks good to me)

individual-it commented:
test change look good to me
The occ command seems to generally not to write to `stdErr`, that might be fixed in an other PR if someone cares

micbar commented:
@sharidas please backport

sharidas commented:
Stable10 PR: https://github.com/owncloud/core/pull/35543

sharidas commented:
Reverting this PR as per the discussion mentioned at https://github.com/owncloud/core/pull/35543#issuecomment-512683143
